src/sinks/datadog_archives.rs

```diff
@@ -0,0 +1,571 @@
+use std::collections::HashSet;
+use std::{collections::BTreeMap, convert::TryFrom};
+
+use bytes::{BufMut, Bytes, BytesMut};
+use chrono::{SecondsFormat, Utc};
+use futures::{stream, FutureExt, SinkExt, StreamExt};
+use global_counter::primitive::exact::CounterU32;
+use lazy_static::lazy_static;
+use rusoto_s3::{PutObjectOutput, S3Client};
+use serde::{Deserialize, Serialize};
+use snafu::Snafu;
+use tower::{util::BoxService, ServiceBuilder};
+use uuid::Uuid;
+
+use vector_core::event::Event;
+
+use crate::config::GenerateConfig;
+use crate::{
+    config::{DataType, SinkConfig, SinkContext},
+    internal_events::TemplateRenderingFailed,
+    rusoto::{AwsAuthentication, RegionOrEndpoint},
+    sinks::{
+        aws_s3::{
+            self, Request, S3CannedAcl, S3RetryLogic, S3ServerSideEncryption, S3Sink,
+            S3StorageClass,
+        },
+        util::{
+            service::Map, BatchSettings, Buffer, Compression, Concurrency, EncodedEvent,
+            PartitionBatchSink, PartitionBuffer, PartitionInnerBuffer, ServiceBuilderExt,
+            TowerRequestConfig, TowerRequestSettings,
+        },
+    },
+    template::Template,
+};
+
+lazy_static! {
+    static ref RESERVED_ATTRIBUTES: HashSet<&'static str> = vec![
+        "_id", "date", "message", "host", "source", "service", "status", "tags", "trace_id",
+        "span_id"
+    ]
+    .into_iter()
+    .collect();
+}
```

Comment:
It's worth considering that every use of `lazy_static` incurs an atomic barrier and we should avoid it on performance grounds. We should also avoid it on the grounds that globals make for complicated to test and understand code. 

Here you use `RESERVED_ATTRIBUTES` in a single place,  the function `encode_event`. Instead of putting a lazily initialized global in the code see if you can't wrap up `encode_event` in a struct that carries the state you need. This'll make testing encoding easier and help us track down and correct any potential performance problems.

Reply:
Thanks for the comments, didn't know about an atomic barrier(wondering if there are lock-free alternatives in Rust),  good to know! I'll try to implement what you  suggest here.

Reply:
If I recall correctly lazy static's implementation is lock free but not wait free. Still, the goal ought to be to avoid any kind of synchronization.

Reply:
I see, thanks, definitely worth getting rid of this.

Reply:
I removed all lazy_static as you suggested.